File: source/modulo_components/test/python/test_component_communication.py
import pytest
from modulo_components.component import Component
from modulo_core.exceptions import CoreError


@pytest.mark.parametrize("minimal_cartesian_input", [[Component, "/topic"]], indirect=True)
@pytest.mark.parametrize("minimal_cartesian_output", [[Component, "/topic", True]], indirect=True)
def test_input_output(ros_exec, random_pose, minimal_cartesian_output, minimal_cartesian_input):
    ros_exec.add_node(minimal_cartesian_input)
    ros_exec.add_node(minimal_cartesian_output)
    ros_exec.spin_until_future_complete(minimal_cartesian_input.received_future, timeout_sec=0.5)
    assert minimal_cartesian_input.received_future.done() and minimal_cartesian_input.received_future.result()
    assert random_pose.get_name() == minimal_cartesian_input.input.get_name()
    assert random_pose.dist(minimal_cartesian_input.input) < 1e-3
    with pytest.raises(CoreError):
        minimal_cartesian_output.publish()


@pytest.mark.parametrize("exception_cartesian_input", [[Component, "/topic"]], indirect=True)
@pytest.mark.parametrize("minimal_cartesian_output", [[Component, "/topic", True]], indirect=True)
def test_exception_input_output(ros_exec, minimal_cartesian_output, exception_cartesian_input):
    ros_exec.add_node(exception_cartesian_input)
    ros_exec.add_node(minimal_cartesian_output)
    ros_exec.spin_until_future_complete(exception_cartesian_input.received_future, timeout_sec=0.5)
    assert exception_cartesian_input.received_future.done() and exception_cartesian_input.received_future.result()


@pytest.mark.parametrize("minimal_cartesian_input", [[Component, "/topic"]], indirect=True)
@pytest.mark.parametrize("minimal_cartesian_output", [[Component, "/topic", False]], indirect=True)
def test_input_output_manual(ros_exec, random_pose, minimal_cartesian_output, minimal_cartesian_input):
    ros_exec.add_node(minimal_cartesian_input)
    ros_exec.add_node(minimal_cartesian_output)
    ros_exec.spin_until_future_complete(minimal_cartesian_input.received_future, timeout_sec=0.5)
    assert not minimal_cartesian_input.received_future.done()
    minimal_cartesian_output.publish()
    ros_exec.spin_until_future_complete(minimal_cartesian_input.received_future, timeout_sec=0.5)
    assert minimal_cartesian_input.received_future.done() and minimal_cartesian_input.received_future.result()
    assert random_pose.get_name() == minimal_cartesian_input.input.get_name()
    assert random_pose.dist(minimal_cartesian_input.input) < 1e-3


@pytest.mark.parametrize("minimal_sensor_input", [[Component, "/topic"]], indirect=True)
@pytest.mark.parametrize("minimal_sensor_output", [[Component, "/topic", False]], indirect=True)
def test_input_output_manual_sensor(ros_exec, random_sensor, minimal_sensor_output, minimal_sensor_input):
    ros_exec.add_node(minimal_sensor_input)
    ros_exec.add_node(minimal_sensor_output)
    ros_exec.spin_until_future_complete(minimal_sensor_input.received_future, timeout_sec=0.5)
    assert not minimal_sensor_input.received_future.done()
    minimal_sensor_output.publish()
    ros_exec.spin_until_future_complete(minimal_sensor_input.received_future, timeout_sec=0.5)
    assert minimal_sensor_input.received_future.done() and minimal_sensor_input.received_future.result()
    for key in random_sensor.get_fields_and_field_types().keys():
        assert getattr(random_sensor, key) == getattr(minimal_sensor_input.input, key)


@pytest.mark.parametrize("minimal_cartesian_input", [[Component, "/topic"]], indirect=True)
@pytest.mark.parametrize("minimal_joint_output", [[Component, "/topic", True]], indirect=True)
def test_input_output_invalid_type(ros_exec, minimal_joint_output, minimal_cartesian_input):
    ros_exec.add_node(minimal_cartesian_input)
    ros_exec.add_node(minimal_joint_output)
    ros_exec.spin_until_future_complete(minimal_cartesian_input.received_future, timeout_sec=0.5)
    assert not minimal_cartesian_input.received_future.done()


@pytest.mark.parametrize("minimal_cartesian_input", [[Component, "/topic"]], indirect=True)
def test_input_output_invalid_msg(ros_exec, make_minimal_invalid_encoded_state_publisher, minimal_cartesian_input):
    invalid_publisher = make_minimal_invalid_encoded_state_publisher("/topic")
    ros_exec.add_node(minimal_cartesian_input)
    ros_exec.add_node(invalid_publisher)
    ros_exec.spin_until_future_complete(minimal_cartesian_input.received_future, timeout_sec=0.5)
    assert not minimal_cartesian_input.received_future.result()


def test_trigger(ros_exec, make_predicates_listener, make_minimal_trigger):
    trigger = make_minimal_trigger(Component)
    listener = make_predicates_listener("/trigger", ["test"])
    ros_exec.add_node(listener)
    ros_exec.add_node(trigger)
    ros_exec.spin_until_future_complete(listener.predicates_future, timeout_sec=0.5)
    assert not listener.predicates_future.done()
    assert not listener.predicate_values["test"]
    trigger.trigger("test")
    ros_exec.spin_until_future_complete(listener.predicates_future, timeout_sec=0.5)
    assert listener.predicates_future.done()
    assert listener.predicate_values["test"]
    assert not trigger.get_predicate("trigger")

File: source/modulo_components/test/cpp/test_component_communication.cpp
#include <gtest/gtest.h>

#include <modulo_utils/testutils/PredicatesListener.hpp>

#include "test_modulo_components/communication_components.hpp"
#include "test_modulo_components/component_public_interfaces.hpp"

using namespace modulo_components;

class ComponentCommunicationTest : public ::testing::Test {
protected:
  void SetUp() override {
    rclcpp::init(0, nullptr);
    exec_ = std::make_shared<rclcpp::executors::SingleThreadedExecutor>();
  }

  void TearDown() override { rclcpp::shutdown(); }

  std::shared_ptr<rclcpp::executors::SingleThreadedExecutor> exec_;
};

TEST_F(ComponentCommunicationTest, InputOutput) {
  auto cartesian_state = state_representation::CartesianState::Random("test");
  auto input_node = std::make_shared<MinimalCartesianInput<Component>>(rclcpp::NodeOptions(), "/topic");
  auto output_node =
      std::make_shared<MinimalCartesianOutput<Component>>(rclcpp::NodeOptions(), "/topic", cartesian_state, true);
  this->exec_->add_node(input_node);
  this->exec_->add_node(output_node);
  auto return_code = this->exec_->spin_until_future_complete(input_node->received_future, 500ms);
  ASSERT_EQ(return_code, rclcpp::FutureReturnCode::SUCCESS);
  EXPECT_EQ(cartesian_state.get_name(), input_node->input->get_name());
  EXPECT_TRUE(cartesian_state.data().isApprox(input_node->input->data()));
  EXPECT_THROW(output_node->publish(), modulo_core::exceptions::CoreException);
}

TEST_F(ComponentCommunicationTest, ExceptionInputOutput) {
  auto cartesian_state = state_representation::CartesianState::Random("test");
  auto input_node = std::make_shared<ExceptionCartesianInput<Component>>(rclcpp::NodeOptions(), "/topic");
  auto output_node =
      std::make_shared<MinimalCartesianOutput<Component>>(rclcpp::NodeOptions(), "/topic", cartesian_state, true);
  this->exec_->add_node(input_node);
  this->exec_->add_node(output_node);
  auto return_code = this->exec_->spin_until_future_complete(input_node->received_future, 500ms);
  ASSERT_EQ(return_code, rclcpp::FutureReturnCode::SUCCESS);
}

TEST_F(ComponentCommunicationTest, InputOutputManual) {
  auto cartesian_state = state_representation::CartesianState::Random("test");
  auto input_node = std::make_shared<MinimalCartesianInput<Component>>(rclcpp::NodeOptions(), "/topic");
  auto output_node =
      std::make_shared<MinimalCartesianOutput<Component>>(rclcpp::NodeOptions(), "/topic", cartesian_state, false);
  this->exec_->add_node(input_node);
  this->exec_->add_node(output_node);
  auto return_code = this->exec_->spin_until_future_complete(input_node->received_future, 500ms);
  ASSERT_EQ(return_code, rclcpp::FutureReturnCode::TIMEOUT);
  output_node->publish();
  return_code = this->exec_->template spin_until_future_complete(input_node->received_future, 500ms);
  ASSERT_EQ(return_code, rclcpp::FutureReturnCode::SUCCESS);
  EXPECT_EQ(cartesian_state.get_name(), input_node->input->get_name());
  EXPECT_TRUE(cartesian_state.data().isApprox(input_node->input->data()));
}

TEST_F(ComponentCommunicationTest, TwistInputOutput) {
  auto twist = std::make_shared<geometry_msgs::msg::Twist>();
  twist->linear.x = 1.0;
  auto input_node = std::make_shared<MinimalTwistInput<Component>>(rclcpp::NodeOptions(), "/topic");
  auto output_node = std::make_shared<MinimalTwistOutput<Component>>(rclcpp::NodeOptions(), "/topic", twist, true);
  this->exec_->add_node(input_node);
  this->exec_->add_node(output_node);
  auto return_code = this->exec_->spin_until_future_complete(input_node->received_future, 500ms);
  ASSERT_EQ(return_code, rclcpp::FutureReturnCode::SUCCESS);
  EXPECT_EQ(twist->linear.x, input_node->input->linear.x);
  EXPECT_THROW(output_node->publish(), modulo_core::exceptions::CoreException);
}

TEST_F(ComponentCommunicationTest, Trigger) {
  auto trigger = std::make_shared<MinimalTrigger<ComponentPublicInterface>>(rclcpp::NodeOptions());
  auto listener =
      std::make_shared<modulo_utils::testutils::PredicatesListener>("/trigger", std::vector<std::string>{"test"});
  this->exec_->add_node(listener);
  this->exec_->add_node(trigger);
  auto result_code = this->exec_->spin_until_future_complete(listener->get_predicate_future(), 500ms);
  ASSERT_EQ(result_code, rclcpp::FutureReturnCode::TIMEOUT);
  EXPECT_FALSE(listener->get_predicate_values().at("test"));
  trigger->trigger();
  result_code = this->exec_->spin_until_future_complete(listener->get_predicate_future(), 500ms);
  ASSERT_EQ(result_code, rclcpp::FutureReturnCode::SUCCESS);
  EXPECT_TRUE(listener->get_predicate_values().at("test"));
  EXPECT_FALSE(trigger->get_predicate("trigger"));
}
